cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -42,13 +42,1440 @@
+
+    RAFT_EXPECTS(
+      chunk_written != -1, "Failed to write to file at offset %lu: %s", file_pos, strerror(errno));
+    RAFT_EXPECTS(chunk_written == static_cast<ssize_t>(chunk_size),
+                 "Incomplete write to file. Expected %zu bytes, wrote %zd at offset %lu",
+                 chunk_size,
+                 chunk_written,
+                 file_pos);
+
+    bytes_remaining -= chunk_size;
+    offset += chunk_size;
+  }
+}
+
+// ACE: Get partition labels for partitioned approach
+template <typename DataT, typename IdxT, typename Accessor>
+void ace_get_partition_labels(
+  raft::resources const& res,
+  raft::mdspan<const DataT, raft::matrix_extent<int64_t>, raft::row_major, Accessor> dataset,
+  raft::host_matrix_view<IdxT, int64_t, raft::row_major> partition_labels,
+  raft::host_matrix_view<IdxT, int64_t, raft::row_major> partition_histogram,
+  uint64_t min_partition_size,
+  double sampling_rate = 0.01)
+{
+  uint64_t dataset_size = dataset.extent(0);
+  uint64_t dataset_dim  = dataset.extent(1);
+  uint64_t labels_size  = partition_labels.extent(0);
+  uint64_t labels_dim   = partition_labels.extent(1);
+  RAFT_EXPECTS(dataset_size == labels_size, "Dataset size must match partition labels extent");
+  uint64_t n_partitions = partition_histogram.extent(0);
+  RAFT_EXPECTS(labels_dim == 2, "Labels must have 2 columns");
+  RAFT_EXPECTS(partition_histogram.extent(1) == 2, "Partition histogram must have 2 columns");
+  cudaStream_t stream = raft::resource::get_cuda_stream(res);
+
+  // Sampling vectors from dataset
+  uint64_t n_samples = dataset_size * sampling_rate;
+  if (n_samples < 100 * n_partitions) { n_samples = 100 * n_partitions; }
+  if (n_samples > dataset_size) { n_samples = dataset_size; }
+  RAFT_LOG_DEBUG("ACE: n_samples: %lu", n_samples);
+
+  auto sample_db = raft::make_host_matrix<float>(n_samples, dataset_dim);
+#pragma omp parallel for
+  for (uint64_t i = 0; i < n_samples; i++) {
+    uint64_t j = i * dataset_size / n_samples;
+    for (uint64_t k = 0; k < dataset_dim; k++) {
+      sample_db(i, k) = static_cast<float>(dataset(j, k));
+    }
+  }
+  auto sample_db_dev = raft::make_device_matrix<float, int64_t>(res, n_samples, dataset_dim);
+  raft::update_device(
+    sample_db_dev.data_handle(), sample_db.data_handle(), sample_db.size(), stream);
+  auto centroids_dev     = raft::make_device_matrix<float, int64_t>(res, n_partitions, dataset_dim);
+  auto sample_labels_dev = raft::make_device_vector<uint32_t, int64_t>(res, n_samples);
+  auto sample_sizes_dev  = raft::make_device_vector<uint32_t, int64_t>(res, n_partitions);
+
+  // K-means: partitioning dataset vectors and compute centroid for each partition
+  // Use balanced k-means with small balancing threshold for more even partition sizes
+  // This shadows build_clusters pattern which doesn't expose balancing threshold.
+  cuvs::cluster::kmeans::balanced_params kmeans_params;
+  kmeans_params.n_iters = 100;
```

Comment:
Do we have evidence that we need such large number of iterations? [Here](https://github.com/facebookresearch/faiss/wiki/FAQ#how-many-training-points-do-i-need-for-k-means) a value of 20 is recommended.

Reply:
Balancing the partitions is important for building high-quality KNN graphs. Thus, the balancing threshold was lower which might require more iterations.